crates/flux-desugar/src/sort_resolver.rs

```diff
@@ -20,8 +22,11 @@ type Result<T = ()> = std::result::Result<T, ErrorGuaranteed>;
 pub(crate) struct SortResolver<'a> {
     sess: &'a FluxSession,
     sort_decls: &'a fhir::SortDecls,
-    generic_params: FxHashMap<Symbol, rustc_span::def_id::DefId>,
+    generic_params: FxHashMap<Symbol, DefId>,
     sort_params: FxHashMap<Symbol, usize>,
+    parent_id: Option<DefId>,
+    /// the sort of `Self` in an `impl` block, None otherwise
+    self_sort: Option<fhir::Sort>,
```

Comment:
Instead of multiple fields I'd have a single field capturing how `Self` should be resolved. We could have an enum like so

```rust
enum SelfRes {
    /// A `Self` parameter in a trait definition.
    Param(DefId),
    /// An alias to another sort, e.g., when used inside an impl block
    Alias(fhir::Sort),
    /// It is not valid to use `Self`, e.g., when used in a free function
    None
}
```

Reply:
yes, this is excellent. Done!